### DISCUSSION TITLE
Improvements to the Swift implementation

### DISCUSSION THREAD

**Discussion by odmir** [PlummersSoftwareLLC/Primes#389] [Improvement suggestions]
Hello! I am looking at the current Swift implementation of the sieve and have some changes in mind that in my machine lead to ~10x speed improvement. The thing is that I consider these changes to be relatively small, and since they don't change the algorithm in any way, I consider them unworthy of a separate solution.

The changes I'm considering are:

1. Changing the Dockerfile to use a more recent version of the language & Ubuntu 20.04 container instead of the current Ubuntu 18.04
2. Change the `Bit count` flag in the README to 8 bits since the implementation is using an array of booleans that each occupy 8 bits in memory, even if only one of those bits is used to represent the value (currently it's set to `unknown`)
3. Changes to `PrimeSwift/solution_1/Sources/PrimeSieveSwift/main.swift`:
	1. Delete the `BitArray` protocol definition (lines 3-7) that is unnecessarily being used as an existential and replace its use in line 33 by the concrete `struct BoolBitArray` that is already implemented
	2. Change `PrimeSieveSwift` from a class to a struct (at line 31) and make the `runSieve` method `mutating` (line 52) to indicate that the method mutates the struct
	3. Add the `bits=8` tag to the program output at line 117
	4. Change the benchmark duration from 10 seconds to 5 at line 125, so the benchmark doesn't take any longer than needed

Then there are more "stylistic" changes that make the code more idiomatic:
1. Replace the use of an implicitly unwrapped optional (IUO) at line 123 by initialising the `sieve` variable to an empty sieve instead (`var sieve: PrimeSieveSwift!` -> `var sieve = PrimeSieveSwift(limit: 0)`)
2. Replace the use of `idx % 2 != 0` by the more descriptive and idiomatic `!idx.isMultiple(of: 2)`
3. Rename the `PrimeSieveSwift` type to `PrimeSieve`
4. Rename the `bits` property to the more descriptive name, `isPrime`
5. Change lines 15-28 in the `BoolBitArray` to be a pair of subscript getter and setter, keeping the indexing implementation the same, and changing every use of the `bits` property to use the subscript operator like `bits[num]` and `bits[num] = false` (at lines 45, 58, 64 and 103), instead of the current `getBit` and `clearBit`

My first question is if I should contribute this as a separate new solution or to the existing one.

The contribution guidelines say this under [faithfulness](https://github.com/PlummersSoftwareLLC/Primes/blob/4ad4c6cf6a503ee4603d362f1e8f2f646e13a5c8/CONTRIBUTING.md#faithfulness): 

> It uses a class to encapsulate the sieve, or an equivalent feature in your language.

So my next question is if my change from a class to a struct falls under "equivalent feature". A C++ class is closer to a Swift struct than a Swift class, so I guess there is no issue here.

Thank you for your attention.

**Comment by tjol** [comment]
As I understand it (just as someone watching this)

 * A pull request with changes that improve the performance will be welcomed
 * Optimisations to existing solutions are preferred to new solutions
 * Cosmetic changes that do not improve performance are not appropriate
 * If there is a suitable construct called 'class' then this must be used. (This came up in relation to C# recently, which I think is similar to Swift in this regard)

**Reply by odmir** [reply]
Thank you for the response.

Interesting, could you point me towards that C# discussion? I am not well versed in C++ or C#, but in Swift a struct is generally a value type while a class is a reference counted entity. A struct can have properties and methods that act on them just like C++ classes. I would like to better understand what is needed for a feature in another language to be considered equivalent to a class mentioned in the guidelines (other than sharing the same name...). Is it that it supports inheritance? Is it that it allows to encapsulate data and methods operating on that data under the same entity? Rust, for example, only has structs and they don't support inheritance, should all rust solutions be marked unfaithful then?

These questions are not directed specifically at you, but at anyone who can share more insight into this. Once again, thank you for your time.

**Reply by tjol** [reply]
https://github.com/PlummersSoftwareLLC/Primes/pull/360#issuecomment-877350820

This is the comment/thread I'm basing my speculation on.

**Reply by odmir** [reply]
Ok, I took a look at it and indeed it seems like a similar issue. I think I will wait for some more perspective from other people closer to the project and that can maybe answer my questions above. At the very least I think that if it needs to be a Swift class to be considered faithful, then, in all fairness, the solutions for the languages that do not have equivalent constructs should be marked unfaithful. Otherwise, "equivalent feature" should be better defined so this kind of ambiguity can be avoided in the future and solutions can be correctly marked as faithful or not.

**Reply by tjol** [reply]
Oh, I don't know. I mean sure, it's arbitrary, but somehow I doubt that it actually matters. Creating a class may be slower than creating a struct, but I'd be surprised if it was so much slower that it's significant, or even reliably measurable, at a couple of thousand iterations per second or however much it is.

**Reply by odmir** [reply]
Unfortunately the Swift compiler is probably having a really hard time having visibility into the class. It doesn't look like an issue with the time it takes to create a class, that should be negligible, but an issue with the optimisations the compiler is able to apply and thus leading to significantly suboptimal code generation. In this case, that small, insignificantly-looking change from class to struct (and adding the `mutating` before the method) results in a speed improvement of >2x. This is very likely missed optimisation opportunities in the compiler and I will be reporting the issue.

**Reply by rbergen** [reply]
The rule is meant to state that for languages that provide a class construct, that is the (only) way to go to have a faithful implementation, but alternatives are accepted where the class construct is not available. I think it is fair feedback on the phrasing that it currently is not fully clear/unambiguous, so I'll move to change the phrasing concerning this.

**Reply by tjol** [reply]
@odmir if accessing the properties of the class is the bottleneck (dynamic message passing overhead or something?), is it possible to just keep a local reference or pointer to the BitArray in the runSieve method instead of accessing the object property every time? (And a local copy of the sieve size)

**Reply by odmir** [reply]
That appears to improve performance very slightly (almost nothing compared to the 2x improvement, though). Something else must be causing such bad code generation. I can't really tell what it is, even after profiling nothing obvious jumps out. Anyway, this is an interesting finding and the compiler should be improved to handle this much better than it currently does.
I will later work on a PR to the solution with some of the other changes, taking into account the feedback here.

**Reply by Gouldsc** [reply]
I would argue that using a class instead of a struct in this context would be a violation of Swift’s language guidelines. Structs are always preferred unless required to facilitate reference semantics, inheritance, framework compatibility, etc. Structs behave very much like traditional classes in Swift, and 10/10 experienced Swift developers porting the original code would be using a struct regardless of the performance win. It would be an “unfaithful implementation” to use a data type in a manner that violates the language guidelines simply because it exists. This isn't an obscure hack to improve performance; it's how the language was designed to be used.

If the intent of the project is to showcase how the same problem is solved in various languages, it would be appropriate to do so in a manner that follows the usage guidelines for that language, instead of requiring nonidiomatic code.

> As a general guideline, prefer structures because they’re easier to reason about, and use classes when they’re appropriate or necessary. In practice, this means most of the custom data types you define will be structures and enumerations.
[Swift Language Guidelines](https://docs.swift.org/swift-book/LanguageGuide/ClassesAndStructures.html)

See also:
[Swift Language Guideline “Choosing between structures and classes”](https://developer.apple.com/documentation/swift/choosing_between_structures_and_classes)

**Comment by odmir** [comment]
Thank you @rbergen for moving this to the right place, I hadn't noticed the discussions section.
